airflow/dag/stringified_dags.py

```diff
@@ -0,0 +1,139 @@
+    'schedule_interval', 'sla', 'execution_timeout'}
+
+_primitive_types = (int, bool, float, str, bytes)
+
+
+def _is_primitive(x):
+    return x is None or isinstance(x, _primitive_types)
+
+
+def _stringify_dag_or_task(x, stringified_dags, is_dag):
+    """Returns a stringified DAG or task."""
+    if is_dag and x.dag_id in stringified_dags:
+        return stringified_dags[x.dag_id]
+
+    # Cast any operators defined in non-airlfow modules to BaseOperator to ensure
+    # unpickle is successful. The downside is that the task will be displayed as
+    # BaseOperator in UI.
+    if not is_dag and not (
+            x.__class__.__module__.startswith('airflow.operators') or
+            x.__class__.__module__.startswith('airflow.contrib.operators')):
+        new_x = object.__new__(models.BaseOperator)
+    else:
+        new_x = object.__new__(x.__class__)
+
+    if is_dag:
+        stringified_dags[x.dag_id] = new_x
+
+    fields_to_keep = _dag_fields_to_keep if is_dag else _task_fields_to_keep
+    for k, v in x.__dict__.items():
+        new_x.__dict__[k] = (copy.deepcopy(v) if k in fields_to_keep else
+                             _stringify(v, stringified_dags))
+    new_x.is_stringified = True
+    return new_x
+
+
+def _stringify(x, stringified_dags):  # pylint: disable=too-many-return-statements
+    """Returns a deep copy of stringified data to make DAGs and tasks picklable.
+
+    Stringifying fields by Depth-First-Search. The order of examing fields is roughly
+    based on their occuring frequency.
+    stringified_dags stores DAGs that are being stringifying for have been stringified,
+    for:
+      (1) preventing deadlock loop caused by task.dag, task._dag, and dag.parent_dag;
+      (2) replacing the fields in (1) with stringified counterparts.
+    """
+    try:
+        if _is_primitive(x):
+            return x
+        elif isinstance(x, dict):
+            return {k: _stringify(v, stringified_dags) for k, v in x.items()}
+        elif isinstance(x, models.DAG):
+            return _stringify_dag_or_task(x, stringified_dags, True)
+        elif isinstance(x, models.BaseOperator):
+            return _stringify_dag_or_task(x, stringified_dags, False)
+        elif callable(x):
+            return _get_python_source(x)
+        elif isinstance(x, list):
+            return [_stringify(v, stringified_dags) for v in x]
+        elif isinstance(x, set):
+            return {_stringify(v, stringified_dags) for v in x}
+        elif isinstance(x, tuple):
+            return tuple([_stringify(v, stringified_dags) for v in x])
+        else:
+            return str(x)
+    except Exception:  # pylint: disable=broad-except
+        logging.warning('Failed to stringify.', exc_info=True)
+        return 'failed_to_stringify'
+
+
+def stringify(x):
+    """Returns a deep copy of stringified data to make DAGs and tasks picklable."""
+    return _stringify(x, {})
+
+
+def _get_python_source(x):
```

Comment:
This function is the same as `airflow.www.utils.get_python_source` isn't it?

Reply:
Yes. But airflow.www.utils.get_python_source has been deprecated since Airflow 1.10.4, so I added it here.

Reply:
Hmm so it has. I think that is a regression and it shouldn't have been removed/should be added back in to master.

Reply:
Added get_python_source back it into airflow/www/utils.py